File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.cpp
// Last updated: 3/26/2026, 1:25:55 PM
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int> st;
        int n = prices.size();
        vector<int> ans(n);
        for(int i =n-1;i>=0;i--){
            while(!st.empty() && prices[i]<prices[st.top()]){
                st.pop();
            }
            ans[i] = prices[i] - (st.empty()?0:prices[st.top()]);
            
            st.push(i);
        }
        return ans;
    }
};

File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.java
// Last updated: 3/26/2026, 1:25:49 PM
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        Stack<Integer> st = new Stack<>();
        for(int i = n-1;i>=0;i--)
        {   int x = prices[i];
            while(!st.isEmpty()&& x<st.peek())
            {
                st.pop();
            }
            if(!st.isEmpty())
            {
                prices[i] -= st.peek();
            }
            st.push(x);
        }
    return prices;    
    }
}